File: Project4/go/task2.go
// implementing a linked list in go making use of some polymorphism to implement the
// data structure 

// Svita Kiran
// CS333
// 10/30/2025

package main
import "fmt"

type Node[T any] struct {
    data T
    next *Node[T]
}

type LinkedList[T any] struct {
    head *Node[T]
    size int
}

// new linked list
func NewLinkedList[T any]() *LinkedList[T] {
    return &LinkedList[T]{}
}

func (l *LinkedList[T]) Push(data T) {
    node := &Node[T]{data: data, next: l.head}
    l.head = node
    l.size++
}

func (l *LinkedList[T]) Pop() *T {
    if l.head == nil {
        return nil
    }
    node := l.head
    l.head = node.next
    l.size--
    return &node.data
}

func (l *LinkedList[T]) Append(data T) {
    node := &Node[T]{data: data}
    if l.head == nil {
        l.head = node
    } else {
        current := l.head
        for current.next != nil {
            current = current.next
        }
        current.next = node
    }
    l.size++
}

func (l *LinkedList[T]) Find(target T, cmp func(a, b T) bool) *T {
    current := l.head
    for current != nil {
        if cmp(current.data, target) {
            return &current.data
        }
        current = current.next
    }
    return nil
}

func (l *LinkedList[T]) Remove(target T, cmp func(a, b T) bool) *T {
    var prev *Node[T]
    current := l.head
    for current != nil {
        if cmp(current.data, target) {
            if prev == nil {
                l.head = current.next
            } else {
                prev.next = current.next
            }
            l.size--
            return &current.data
        }
        prev = current
        current = current.next
    }
    return nil
}

func (l *LinkedList[T]) Map(f func(data T)) {
    current := l.head
    for current != nil {
        f(current.data)
        current = current.next
    }
}

func (l *LinkedList[T]) Size() int {
    return l.size
}

// clear the list
func (l *LinkedList[T]) Clear() {
    l.head = nil
    l.size = 0
}


// test program
func main() {
	intList := NewLinkedList[int]()
    for i := 0; i < 5; i++ {
        intList.Push(i)
    }
    fmt.Println("integer list:")
    intList.Map(func(i int) { fmt.Println(i) })

	strList := NewLinkedList[string]()
    words := []string{"car", "bike", "train"}
    for _, w := range words {
        strList.Append(w)
    }
    fmt.Println("\nstring list:")
    strList.Map(func(s string) { fmt.Println(s) })

	found := intList.Find(3, func(a, b int) bool { return a == b })
    if found != nil {
        fmt.Println("\nfound:", *found)
    }

	removed := strList.Remove("bike", func(a, b string) bool { return a == b })
    if removed != nil {
        fmt.Println("\nremoved:", *removed)
    }
    fmt.Println("\nafter removal:")
    strList.Map(func(s string) { fmt.Println(s) })
}

File: Project4/haskell/task2.hs
{- implementing a linked list in go making use of some polymorphism to implement the
data structure 

Svita Kiran
CS333
10/30/2025
-}

data LinkedList a = Empty | Node a (LinkedList a)
    deriving Show

llCreate :: LinkedList a
llCreate = Empty

llPush :: a -> LinkedList a -> LinkedList a
llPush x xs = Node x xs

llPop :: LinkedList a -> (Maybe a, LinkedList a)
llPop Empty = (Nothing, Empty)
llPop (Node x xs) = (Just x, xs)

llAppend :: a -> LinkedList a -> LinkedList a
llAppend x Empty = Node x Empty
llAppend x (Node y ys) = Node y (llAppend x ys)

llFind :: (a -> Bool) -> LinkedList a -> Maybe a
llFind _ Empty = Nothing
llFind pred (Node x xs)
    | pred x    = Just x
    | otherwise = llFind pred xs

llRemove :: (a -> Bool) -> LinkedList a -> (Maybe a, LinkedList a)
llRemove _ Empty = (Nothing, Empty)
llRemove pred (Node x xs)
    | pred x    = (Just x, xs)
    | otherwise = let (res, rest) = llRemove pred xs
                  in (res, Node x rest)

llSize :: LinkedList a -> Int
llSize Empty = 0
llSize (Node _ xs) = 1 + llSize xs

llMap :: (a -> b) -> LinkedList a -> LinkedList b
llMap _ Empty = Empty
llMap f (Node x xs) = Node (f x) (llMap f xs)

llClear :: LinkedList a -> LinkedList a
llClear _ = Empty


-- test program
main :: IO ()
main = do
    let list1 = llPush 1 $ llPush 2 $ llPush 3 llCreate
    print list1
    print $ llSize list1
    print $ llFind (==2) list1

    let list2 = llPush "car" $ llPush "bike" llCreate
    print list2
    print $ llFind (=="car") list2

    let list2Appended = llAppend "train" list2
    print $ llMap (++"!") list2Appended